lib/net/x86-64-linux/Net.v3

```diff
@@ -150,59 +147,229 @@ def socketToSockAddr(socket: Socket, server: bool) -> Array<byte> {
+			fd = Net.socket(family, sType, 0 /* protocol */);
 		} else {
 			fd = sockfd;
 		}
 	}
 
+	def setBound(bound: bool) {
+		isBound = bound;
+	}
+
+	def setListening(listening: bool) {
+		isListening = listening;
+	}
+
+	def setConnected(connected: bool) {
+		isConnected = connected;
+	}
+
+	def getLocalAddress() -> InetAddress;
+
+	def getSockAddr() -> Array<byte>;
+
 	def setLocalPort(local: int) {
 		localPort = local;
 	}
 
-	// Read a message into buffer {buf} from the socket. Return how many bytes
-	// were read into the buffer.
-	def read(buf: Array<byte>) -> int {
-		return System.fileReadK(fd, buf, 0, buf.length);
+	def isUdp() -> bool {
+		return socketType == SocketType.Udp;
+	}
+
+	// Receive a message into buffer {buf} from the socket. Return how many bytes
+	// were read into the buffer. The receive will fail on a standard
+	// connectionless UDP socket.
+	def recv(buf: Array<byte>) -> int {
+		if (!isUdp() || isConnected) {
+			return System.fileReadK(fd, buf, 0, buf.length);
+		} else {
+			// socketType == SocketType.Udp
+			System.error("Error", "Called recv() on an unconnected UDP socket! Please use recvfrom() instead.");
+			return -1;
+		}
 	}
 
-	// Write a message {msg} into the socket.
-	def write(msg: string) {
-		System.fileWriteK(fd, msg, 0, msg.length);
+	// Receive a message into buffer {buf} from a remote host. Users can provide
+	// an instance of {packet} which will be overridden with information about
+	// the remote host. If no {packet} is provided, a new one is allocated and
+	// returned. Returns the number of bytes read into {buf} and information
+	// about the remote host for future communication. The receive will fail for
+	// a TCP socket.
+	def recvfrom(buf: Array<byte>, packet: UdpPacket) -> (int, UdpPacket) {
+		if (!isUdp()) {
+			System.error("Error", "Called recvfrom() on a TCP socket!");
+		}
+
+		if (packet == null) {
+			packet = UdpPacket.new(Net.ANY_V4, -1);
+		}
+
+		var retSockAddr = Array<byte>.new(LinuxConst.SOCKADDR_STORAGE_SIZE);
+		var retSock = Net.recvfrom(fd, buf, retSockAddr);
+		def retSockAddrSize = retSock.1.x;
+
+		retSockAddr = Arrays.range(retSockAddr, 0, retSockAddrSize);
+		def retPort = (int.view(retSockAddr[2]) << 8) | int.view(retSockAddr[3]);
+
+		match (retSockAddrSize) {
+			LinuxConst.SOCKADDR_V4_SIZE => {
+				def retAddr = Net.rangeToInetAddress(retSockAddr[4 ... 8]);
+				packet.addr = retAddr;
+				packet.port = retPort;
+				return (retSock.0, packet);
+			}
+			LinuxConst.SOCKADDR_V6_SIZE => {
+				def retAddr = Net.rangeToInetAddress(retSockAddr[8 ... 24]);
+				packet.addr = retAddr;
+				packet.port = retPort;
+				return (retSock.0, packet);
+			}
+			_ => {
+				System.error("Error", "Unexpected size for returned sockaddr!");
+			}
+		}
+
+		return (-1, null);
+	}
+
+	// Send a message {msg} into the socket. The send will fail on a standard UDP
+	// connectionless socket.
+	def send(msg: string) {
```

Comment:
Same here, I think `Range<byte>` would be more general.

I also have plans of making `string` into a proper immutable array of Unicode code points.